packages/material-ui/src/ClickAwayListener/ClickAwayListener.test.js

```diff
@@ -56,6 +59,84 @@ describe('<ClickAwayListener />', () => {
 
       document.body.removeEventListener('click', preventDefault);
     });
+
+    it('should not be called when clicking inside a portaled element', () => {
+      const handleClickAway = spy();
+      const { getByText } = render(
+        <ClickAwayListener onClickAway={handleClickAway}>
+          <div>
+            <Portal>
+              <span>Inside a portal</span>
+            </Portal>
+          </div>
+        </ClickAwayListener>,
+      );
+
+      fireEvent.click(getByText('Inside a portal'));
+      expect(handleClickAway.callCount).to.equal(0);
+    });
+
+    it('should be called when clicking inside a portaled element and `disableReactTree` is `true`', () => {
+      const handleClickAway = spy();
+      const { getByText } = render(
+        <ClickAwayListener onClickAway={handleClickAway} disableReactTree>
+          <div>
+            <Portal>
+              <span>Inside a portal</span>
+            </Portal>
+          </div>
+        </ClickAwayListener>,
+      );
+
+      fireEvent.click(getByText('Inside a portal'));
+      expect(handleClickAway.callCount).to.equal(1);
+    });
+
+    it('should not be called even if the event propagation is stopped', () => {
+      const handleClickAway = spy();
+      const { getByText } = render(
+        <ClickAwayListener onClickAway={handleClickAway} disableReactTree>
+          <div>
+            <div
+              onClick={(event) => {
+                event.stopPropagation();
+              }}
+            >
+              Outside a portal
+            </div>
+            <Portal>
+              <span
+                onClick={(event) => {
+                  event.stopPropagation();
+                }}
+              >
+                Stop inside a portal
+              </span>
+            </Portal>
+            <Portal>
+              <span
+                onClick={(event) => {
+                  event.stopPropagation();
+                  event.nativeEvent.stopImmediatePropagation();
+                }}
+              >
+                Stop all inside a portal
+              </span>
+            </Portal>
+          </div>
+        </ClickAwayListener>,
+      );
+
+      fireEvent.click(getByText('Outside a portal'));
+      expect(handleClickAway.callCount).to.equal(0);
+
+      fireEvent.click(getByText('Stop all inside a portal'));
+      expect(handleClickAway.callCount).to.equal(0);
+
+      fireEvent.click(getByText('Stop inside a portal'));
+      // True-negative, we don't have enough information to do otherwise.
+      expect(handleClickAway.callCount).to.equal(1);
```

Comment:
@NMinhNguyen This is undesired behavior, no? Since the target handler stops the propagation we shouldn't register a click to begin with?

Reply:
I think this particular test was written by @oliviertassinari - https://github.com/mui-org/material-ui/commit/1a306e26c41961b31b0165e7ceb06694c9123a51 who may be able to provide more clarity on this. Sorry, it's been a while, I don't quite remember the details now.

Reply:
Ah check this actually https://github.com/mui-org/material-ui/pull/20406#discussion_r403537212

Reply:
I'm not asking why this is happening bug if this is desired or undesired behavior. 

Seems to me that having the React event system work closer to how the DOM event system work is desired.

Reply:
Right. I'm personally not sure, hence why I raised that question in the first place, and I let @oliviertassinari make a call. I'll probably have to defer to Olivier on your question though.

Reply:
Considering the context Sebastian is coming from, I would suspect a regression from the next version of React. This test case was meant to document a structural limitation of the approach. In the test case, the event.stopPropagation call should be respected, it seems to be ignored.

Reply:
> This test case was meant to document a structural limitation of the solution to the problem developed here.

What does that mean?

> In the test case, the event.stopPropagation call should be respected by React but it seems to be ignored, not preventing the event to bubble up.

So this is indeed undesired behavior?

Reply:
> What does that mean?

I'm on my phone, I would need to test more closely what's going on. Now, what it meant is that the test should be failing. It's not meant to work, in theory, it can't work. If it does, React has changed some of the previous constraint. The first assumption I have, that I would need to test and that would explain the failing test is that React broke the React stop propagation logic when it involves a portal.

Reply:
Please ignore what react is doing, what the DOM is doing, what any of the implementation is doing. Do you think that the behavior that is expected in the test is desired behavior?

Reply:
Had some further thoughts actually. Given it says "True-negative" in the comment, I'm inclined to say it's undesirable.

Reply:
@eps1lon To further clarify what I had in the back of my mind during the implementation: 

- I shouldn't have used: "True-negative", it wasn't the right description, backward actually, sorry for the confusion. It should have been "False-positive", "False" as the behavior is wrong, and "positive" as it raises when it shouldn't.
- The outcome was a hard limitation of the solution we took. We needed React to have a different behavior to have a True-negative (the desired outcome).

Having a closer look at #22105, it seems that the new behavior of React is to do `event.nativeEvent.stopImmediatePropagation();` when `event.stopPropagation();` is called on the synthetic event, which is perfect for us, we no longer have a "False-positive" :D.

Reply:
Yeah, personally I try to avoid "true-negative" or "false-positive" since they are meaningless unless you specific to what pool "true" and "negative" belong to.

I think it makes more sense to use the following mapping:
- behavior in a failing test that shouldn't fail:  unexpected and undesired behavior
- behavior in a failing test that should fail: unexpected and desired
- behavior in a passing test that shouldn't pass: expected and undesired behavior
- behavior in a passing test that should pass: expected and desired behavior

(un)expected behavior is basically what you write in your test and (un)desired behavior is what your (un)expected behavior should eventually become.

So here we had expected but undesired behavior which is unexpected but desired behavior in `react@next` and will become expected and desired behavior in https://github.com/mui-org/material-ui/pull/22105#discussion_r467308235

Reply:
Sounds about right 👌

Reply:
For posterity, https://reactjs.org/blog/2020/08/10/react-v17-rc.html#fixing-potential-issues explains how `event.stopPropagation()` works in React 17.